site/content/docs/user/exposing-applications.md

```diff
@@ -0,0 +1,243 @@
+Type:                     NodePort
+IP:                       10.96.59.230
+Port:                     <unset>  8080/TCP
+TargetPort:               8080/TCP
+NodePort:                 <unset>  30039/TCP
+Endpoints:                10.36.0.1:8080,10.44.0.1:8080
+Session Affinity:         None
+External Traffic Policy:  Cluster
+Events:                   <none>
+```
+
+
+### Nginx ingress controller
+
+> Ingress is actually NOT a type of service. Instead, it sits in front of multiple services and act as a “smart router” or entrypoint into your cluster.
+
+TODO
+
+## Accesing the application
+
+For some of this methods we'll need the cluster IP addresses
+
+```
+kubectl get nodes -o wide
+
+NAME                 STATUS   ROLES    AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                                  KERNEL-VERSION     CONTAINER-RUNTIME
+kind-control-plane   Ready    master   7m1s    v1.14.1   172.17.0.4    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+kind-worker          Ready    <none>   5m58s   v1.14.1   172.17.0.2    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+kind-worker2         Ready    <none>   5m59s   v1.14.1   172.17.0.3    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+```
+
+### Kubectl port-forward
+
+This is the most simple one and well documented
+
+https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/
+
+### Port forwarding with socat
+
+Docker allows to port map containers ports to our localhost, the problem is that this port mapping can be configured only when we create the container. If we want to map new port we have to recreate the container.
+
+We can leverage this functionality to create one container that allows to forward ports, this container will forward the new ports mapped to the corresponding ports on the cluster.
+
+This process is well documented on different places, ie.
+
+https://sosedoff.com/2018/04/25/expose-docker-ports.html
+
+https://andrewwilkinson.wordpress.com/2017/11/22/exposing-docker-ports-after-the-fact/
+
+
+
+
+### Socks Proxy
+
+> A SOCKS server is a general purpose proxy server that establishes a TCP connection to another server on behalf of a client, then routes all the traffic back and forth between the client and the server. It works for any kind of network protocol on any port
+
+Instead of having to recreate the port mapping constantly we can create a socks proxy inside the docker network and use it to connect to the exposed applications.
+
+SOCKS proxy can be used with any web browser, just have to configure your application to use it and use your service IP and Port target directly.
+
+Create a container with a SOCKS proxy and forward the port
+
+```sh
+docker run --rm -d --privileged --name socks5 -p 1080:1080 aojea/socks5
```

Comment:
Probably don't need such a big privilege? May give it NET_ADMIN permission. Using --cap-add NET_ADMIN

(But I need to test it later.)

Reply:
> Probably don't need such a big privilege? May give it NET_ADMIN permission. Using --cap-add NET_ADMIN
> 
> (But I need to test it later.)

 Tested, as I said we can just give it `NET_ADMIN` permission only.  :smile: